### PR TITLE
[ti_opencti] Fix processing of externalReferences

### LINKED ISSUE #8457: ti_opencti parsing failures lead to dropped events
As below,

```
09:22:25.553
elastic_agent.filebeat
[elastic_agent.filebeat][warn] Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Date(2023, time.November, 9, 23, 22, 22, 853987964, time.Local), Meta:{"input_id":"cel-opencti-83b801b7-46d1-4b1b-beed-dcbf8c0fa273","raw_index":"logs-ti_opencti.indicator-default","stream_id":"cel-ti_opencti.indicator-83b801b7-46d1-4b1b-beed-dcbf8c0fa273"}, Fields:{"_conf":{"url":"https://opencti.example.tld"},"agent":{"ephemeral_id":"d03475d0-61e6-4c85-8d6f-11d51dd04997","id":"7d3d7879-7cd6-4352-9c2b-186499853bc2","name":"jump01-syd","type":"filebeat","version":"8.10.4"},"cloud":{"availability_zone":"","instance":{"id":"","name":""},"provider":"hetzner","region":"","service":{"name":"Cloud"}},"confidence":0,"created":"2023-11-02T00:17:00.295Z","createdBy":{"identity_class":"organization","name":"Stopforumspam"},"data_stream":{"dataset":"ti_opencti.indicator","namespace":"default","type":"logs"},"description":"Stopforumspam","ecs":{"version":"8.0.0"},"elastic_agent":{"id":"7d3d7879-7cd6-4352-9c2b-186499853bc2","snapshot":false,"version":"8.10.4"},"event":{"dataset":"ti_opencti.indicator","original":{"confidence":0,"created":"2023-11-02T00:17:00.295Z","createdBy":{"identity_class":"organization","name":"Stopforumspam"},"description":"Stopforumspam","externalReferences":{"edges":[{"node":{"description":"Stopforumspam feed URL","source_name":"stopforumspam","url":"https://www.stopforumspam.com/downloads/toxic_domains_whole_filtered_50000.txt"}}]},"id":"fcfa872e-a8b6-4525-847e-f3c756b70035","is_inferred":false,"killChainPhases":{"edges":null},"lang":"en","modified":"2023-11-09T23:22:20.586Z","name":"freelifetimexxxdates.com","objectLabel":{"edges":[{"node":{"value":"spam"}}]},"objectMarking":{"edges":null},"observables":{"edges":[{"node":{"entity_type":"Domain-Name","id":"cc34949a-5a6f-4595-afec-c3bf98c62a7d","observable_value":"freelifetimexxxdates.com","standard_id":"domain-name--726e8863-8941-5a1b-b345-1f0131902233","value":"freelifetimexxxdates.com"}}],"pageInfo":{"globalCount":1}},"pattern":"[domain-name:value = 'freelifetimexxxdates.com']","pattern_type":"stix","revoked":false,"standard_id":"indicator--08a7e875-2ce4-50ab-a8de-2915addd93c4","valid_from":"2023-11-09T23:22:19.426Z","valid_until":"2024-11-08T23:22:19.426Z","x_opencti_detection":false,"x_opencti_main_observable_type":"Domain-Name","x_opencti_score":60}},"externalReferences":{"edges":[{"node":{"description":"Stopforumspam feed URL","source_name":"stopforumspam","url":"https://www.stopforumspam.com/downloads/toxic_domains_whole_filtered_50000.txt"}}]},"id":"fcfa872e-a8b6-4525-847e-f3c756b70035","input":{"type":"cel"},"is_inferred":false,"killChainPhases":{"edges":null},"lang":"en","modified":"2023-11-09T23:22:20.586Z","name":"freelifetimexxxdates.com","objectLabel":{"edges":[{"node":{"value":"spam"}}]},"objectMarking":{"edges":null},"observables":{"edges":[{"node":{"entity_type":"Domain-Name","id":"cc34949a-5a6f-4595-afec-c3bf98c62a7d","observable_value":"freelifetimexxxdates.com","standard_id":"domain-name--726e8863-8941-5a1b-b345-1f0131902233","value":"freelifetimexxxdates.com"}}],"pageInfo":{"globalCount":1}},"pattern":"[domain-name:value = 'freelifetimexxxdates.com']","pattern_type":"stix","revoked":false,"standard_id":"indicator--08a7e875-2ce4-50ab-a8de-2915addd93c4","tags":["preserve_original_event","forwarded","opencti-indicator"],"valid_from":"2023-11-09T23:22:19.426Z","valid_until":"2024-11-08T23:22:19.426Z","x_opencti_detection":false,"x_opencti_main_observable_type":"Domain-Name","x_opencti_score":60}, Private:interface {}(nil), TimeSeries:false}, Flags:0x1, Cache:publisher.EventCache{m:mapstr.M(nil)}} (status=400): {"type":"document_parsing_exception","reason":"[1:3100] object mapping for [opencti.indicator.external_reference] tried to parse field [null] as object, but found a concrete value"}, dropping event!
09:22:25.554
elastic_agent.filebeat
[elastic_agent.filebeat][warn] Cannot index event publisher.Event{Content:beat.Event{Timestamp:time.Date(2023, time.November, 9, 23, 22, 22, 854015277, time.Local), Meta:{"input_id":"cel-opencti-83b801b7-46d1-4b1b-beed-dcbf8c0fa273","raw_index":"logs-ti_opencti.indicator-default","stream_id":"cel-ti_opencti.indicator-83b801b7-46d1-4b1b-beed-dcbf8c0fa273"}, Fields:{"_conf":{"url":"https://opencti.example.tld"},"agent":{"ephemeral_id":"d03475d0-61e6-4c85-8d6f-11d51dd04997","id":"7d3d7879-7cd6-4352-9c2b-186499853bc2","name":"jump01-syd","type":"filebeat","version":"8.10.4"},"cloud":{"availability_zone":"","instance":{"id":"","name":""},"provider":"hetzner","region":"","service":{"name":"Cloud"}},"confidence":0,"created":"2023-11-02T00:17:04.260Z","createdBy":{"identity_class":"organization","name":"Stopforumspam"},"data_stream":{"dataset":"ti_opencti.indicator","namespace":"default","type":"logs"},"description":"Stopforumspam","ecs":{"version":"8.0.0"},"elastic_agent":{"id":"7d3d7879-7cd6-4352-9c2b-186499853bc2","snapshot":false,"version":"8.10.4"},"event":{"dataset":"ti_opencti.indicator","original":{"confidence":0,"created":"2023-11-02T00:17:04.260Z","createdBy":{"identity_class":"organization","name":"Stopforumspam"},"description":"Stopforumspam","externalReferences":{"edges":[{"node":{"description":"Stopforumspam feed URL","source_name":"stopforumspam","url":"https://www.stopforumspam.com/downloads/toxic_domains_whole_filtered_50000.txt"}}]},"id":"43efd186-23dd-438a-b963-97f9ef84a9d9","is_inferred":false,"killChainPhases":{"edges":null},"lang":"en","modified":"2023-11-09T23:22:21.423Z","name":"freeliveadultcams.com","objectLabel":{"edges":[{"node":{"value":"spam"}}]},"objectMarking":{"edges":null},"observables":{"edges":[{"node":{"entity_type":"Domain-Name","id":"4c1d51fc-861c-41fc-9f50-fe11a142b86a","observable_value":"freeliveadultcams.com","standard_id":"domain-name--95b63b1f-4425-5382-8814-d4a740057fc7","value":"freeliveadultcams.com"}}],"pageInfo":{"globalCount":1}},"pattern":"[domain-name:value = 'freeliveadultcams.com']","pattern_type":"stix","revoked":false,"standard_id":"indicator--7b57ce8f-5ee2-5689-ac2c-ec27dc68f8b0","valid_from":"2023-11-09T23:22:20.422Z","valid_until":"2024-11-08T23:22:20.422Z","x_opencti_detection":false,"x_opencti_main_observable_type":"Domain-Name","x_opencti_score":60}},"externalReferences":{"edges":[{"node":{"description":"Stopforumspam feed URL","source_name":"stopforumspam","url":"https://www.stopforumspam.com/downloads/toxic_domains_whole_filtered_50000.txt"}}]},"id":"43efd186-23dd-438a-b963-97f9ef84a9d9","input":{"type":"cel"},"is_inferred":false,"killChainPhases":{"edges":null},"lang":"en","modified":"2023-11-09T23:22:21.423Z","name":"freeliveadultcams.com","objectLabel":{"edges":[{"node":{"value":"spam"}}]},"objectMarking":{"edges":null},"observables":{"edges":[{"node":{"entity_type":"Domain-Name","id":"4c1d51fc-861c-41fc-9f50-fe11a142b86a","observable_value":"freeliveadultcams.com","standard_id":"domain-name--95b63b1f-4425-5382-8814-d4a740057fc7","value":"freeliveadultcams.com"}}],"pageInfo":{"globalCount":1}},"pattern":"[domain-name:value = 'freeliveadultcams.com']","pattern_type":"stix","revoked":false,"standard_id":"indicator--7b57ce8f-5ee2-5689-ac2c-ec27dc68f8b0","tags":["preserve_original_event","forwarded","opencti-indicator"],"valid_from":"2023-11-09T23:22:20.422Z","valid_until":"2024-11-08T23:22:20.422Z","x_opencti_detection":false,"x_opencti_main_observable_type":"Domain-Name","x_opencti_score":60}, Private:(*cursor.updateOp)(0xc00125d840), TimeSeries:false}, Flags:0x1, Cache:publisher.EventCache{m:mapstr.M(nil)}} (status=400): {"type":"document_parsing_exception","reason":"[1:3073] object mapping for [opencti.indicator.external_reference] tried to parse field [null] as object, but found a concrete value"}, dropping event!

```

### PR DESCRIPTION
**Pull request by chrisberkhout**
## Proposed commit message

```
[ti_opencti] Fix processing of externalReferences (#8556)

The contents of externalReferences.edges were being processed with the
append processor and mistakenly flattened into scalar values. To fix
that they are now handled in Painless, as objects.

Additional checks were added for collection fields that may be null,
after some were observed in example data added to test the
externalReferences fix.
```

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [x] Merge of https://github.com/elastic/integrations/pull/8428

## How to test this PR locally

```
cd packages/ti_opencti
elastic-package stack up -d
elastic-package test -v
```

## Related issues

- Closes #8457

### CODE REVIEW

File: packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -166,19 +170,35 @@ processors:
...
-      processor:
-        append:
-          field: opencti.indicator.external_reference
-          value: "{{{_ingest._value.node}}}"
+  - script:
+      description: Move external reference nodes to their target location
+      lang: painless
+      source: |
+        ArrayList edges = ctx.externalReferences.edges;
+        if (edges != null) {
```
Comment (efd6) [inline packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml:186]: Style only: consider inverting the condition and returning early to reduce indentation.

File: packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      "externalReferences": {
+        "edges": [
+          {
+            "node": {
+              "description": "Stopforumspam feed URL",
+              "source_name": "stopforumspam",
+              "url": "https://www.stopforumspam.com/downloads/toxic_domains_whole_filtered_50000.txt"
+            }
+          }
+        ]
```
Comment (efd6) [inline packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json:20]: Do we see documents with more than one edge? Can we have test for that case?

File: packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      },
+      "objectMarking": {
+        "edges": null
+      },
+      "observables": {
+        "edges": [
+          {
```
Comment (efd6) [inline packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json:44]: Is this something that we want to do for this as well?

**Review by efd6 — COMMENTED** [review]
Changelog and manifest updates?

File: packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml (hunk 1 above)
Comment (chrisberkhout) [reply packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml]: Done.

File: packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json (hunk 2 above)
Comment (chrisberkhout) [reply packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json]: It's probably very rare in the data, but the schema allows it. I added a second edge in the test case and fixed a null handling issue that exposed.

File: packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json (hunk 3 above)
Comment (chrisberkhout) [reply packages/ti_opencti/data_stream/indicator/_dev/test/pipeline/test-domain-name-with-external-reference.json:52]: The observables do already get the treatment added here for external references.

**Comment by chrisberkhout** [comment]
> Changelog and manifest updates?

Now done. I should have opened this as a draft, since I wanted to wait until community PR https://github.com/elastic/integrations/pull/8428 was merged.

File: packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -166,19 +170,42 @@ processors:
...
+  - script:
+      description: Move external reference nodes to their target location, without nulls
+      lang: painless
+      source: |
+        ArrayList edges = ctx.externalReferences.edges;
+        if (edges == null) {
+          return;
+        }
+        for (int i = 0; i < edges.length; i++) {
+          if (!ctx.opencti?.indicator?.containsKey('external_reference')) {
```
Comment (efd6) [inline packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml:190]: This is not safe; the `?.` operator stops the derefs being unsafe, but then we get to a look up for the method on `null`, and bang!
```suggestion
          if (ctx.opencti?.indicator != null && !ctx.opencti.indicator.containsKey('external_reference')) {
```

Comment (chrisberkhout) [reply packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml:190]: Here the last null safe operator replaces a plain `.containsKey()` method call. My rule of thumb is that `?.` will make the thing immediately to its right safe - either a field access or a method call.

This script returns "safely got a null":
```
POST _scripts/painless/_execute
{
  "script": {
    "params": {
      "key": null
    },
    "source": """
      if (params.key?.nonKey?.nonMethod("arg")?.nonMethod2("arg") == null) {
        return "safely got a null";
      }
    """
  }
}
```

Comment (chrisberkhout) [reply packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml]: Actually, I see there's a problem. The `...?.containsKey('external_reference')` is okay, but `null` as the result of an `if` condition is not okay. So I fixed that and other occurrences like this
```
if (ctx.opencti?.indicator?.containsKey('external_reference') == true) {
  ...
}
```

Comment (efd6) [reply packages/ti_opencti/data_stream/indicator/elasticsearch/ingest_pipeline/default.yml]: Thanks, that's very informative; I've learned something new.

**Review by efd6 — APPROVED** [review]
Thanks

**Pull request merged**